walkdir.py

```diff
@@ -202,11 +202,32 @@ def onloop(dirpath):
                     continue
         yield dir_entry
 
+
+def handle_dirsymlink_as_file(walk_iter):
+    """Handle symlink to directories as files during the walk.
+
+       By default os.walk insert symlink to directories to the subdirs list.
+       This function move symlink to directories from the subdirs list to the file list.
+
+       NOTE: This function is not needed when using the followlinks option.
+       when following links the dirsymlinks will be walked when entering the linked dir.
+    """
+    for dir_entry in walk_iter:
+        dirpath = dir_entry[0]
+        subdirs = dir_entry[1]
+        files = dir_entry[2]
+        for subdir in subdirs:
```

Comment:
Modifying the file and subdir lists on each iteration makes the overall approach O(N^2), so may be a problem when processing large directories. Accordingly, I suggest building up two lists in the loop, and then applying them in a single step after the loop:

    files.extend(symlinked_subdirs)
    subdirs[:] = direct_subdirs

This does use a bit more memory, but I'd expect to see problems from the O(N^2) algorithm before we saw them from effectively duplicating the subdirs list. (The memory duplication could be avoided, but it needs a stat cache as described in [1] to avoid doubling the number of stat calls)

[1] https://bitbucket.org/ncoghlan/walkdir/issues/20/use-an-lstat-function-with-implicit-time